### PR TITLE
Calendars: Astro functions

### PR DESCRIPTION
**Pull request by gzotti**
### Description

This adds the ~100 astro functions required for astronomically guided calendars in the second half of the book.
I have tried to use Stellarium's functions and data types (e.g. StelLocation) where applicable. 

So far I cannot test those functions. Whenever a StelCore is needed, the test functions fail. So it's unlikely that this is all error-free...

Fixes # (issue)

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] This change requires a documentation update

### How Has This Been Tested?

**Test Configuration**:
* Operating system: Win10
* Graphics Card: irrelevant

## Checklist:

- [x] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (header file)
- [x] I have updated the respective chapter in the Stellarium User Guide
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by alex-w** [comment]
> So far I cannot test those functions. Whenever a StelCore is needed, the test functions fail. So it's unlikely that this is all error-free...

This is correct behaviour, because StelApp is singleton. Please check DeltaT test for R&D algorithm.

**Comment by gzotti** [comment]
> > So far I cannot test those functions. Whenever a StelCore is needed, the test functions fail. So it's unlikely that this is all error-free...
> 
> This is correct behaviour, because StelApp is singleton. 

Yes, the CI tests run in debug mode and test for this, and therefore fail as well...

> Please check DeltaT test for R&D algorithm.

I have fixed one error from the latest official error list from R&D. Else, in the program, I use the core's currently used DeltaT algorithm. R&D's version is just another reformulation of Meeus AA1 (1991), and can actually be removed.

Not sure how to test and continue with these things now. Just implement further calendars on top, and see if they work in the program? (CI/tests won't work for those!)

One idea of a future benefit from implementing all this would be scripting accessibility. Just make slots from them? This could also allow at least manual/semi-automatic testing.

**Comment by alex-w** [comment]
> > Please check DeltaT test for R&D algorithm.
> 
> I have fixed one error from the latest official error list from R&D. Else, in the program, I use the core's currently used DeltaT algorithm. R&D's version is just another reformulation of Meeus AA1 (1991), and can actually be removed.

But R&D solution has fits of expanded time range in comparison to AA1.

> Not sure how to test and continue with these things now. Just implement further calendars on top, and see if they work in the program? (CI/tests won't work for those!)

Probably code of methods should be reformed or omitted StelApp/StelCore-related methods from unit testing.

> One idea of a future benefit from implementing all this would be scripting accessibility. Just make slots from them? This could also allow at least manual/semi-automatic testing.

Well, at the moment this is best solution IMHO.

**Comment by alex-w** [comment]
@gzotti could you check isLeap method for Gregorian calendar?

**Comment by gzotti** [comment]
I am stuck with the scripting interface. Just simple QVector<int> which should be converted to JavaScript Arrays. The other useful object to make scriptable in this context would be StelLocation. The current situation is barely enough for scriptable debugging output. @alex-w any idea how to make these?

**Comment by alex-w** [comment]
Oops, sorry, I missed this question. :(

> I am stuck with the scripting interface. Just simple QVector which should be converted to JavaScript Arrays. 

Why? Do you meant `QVector::toList()`? 

> The other useful object to make scriptable in this context would be StelLocation. 

See [setObserverLocation](https://stellarium.org/doc/0.21/classStelMainScriptAPI.html#aa6655a5dc00083085c3fc12646536231) in the scripting's core and both getObserverLocation methods in same "class".

> The current situation is barely enough for scriptable debugging output. @alex-w any idea how to make these?

No idea right now, but... see how Vec3d/Vec2d types was added into scripting engine

**Comment by gzotti** [comment]
I played for a while with additional functions wrapping the QVector with .toList(). But now I found the right conversion/registration (see StelScriptMgr), so this part is solved, and most calendar functions are scriptable :-). 

Thanks for the hint about setObserverLocation, this may help. And some sleep before doing this ;-)

**Comment by alex-w** [comment]
Maybe adding context for calendar names in the GUI will be good idea?

File: src/core/StelUtils.cpp [hunk 1]
```diff
@@ -1776,7 +1776,7 @@ double getDeltaTByReingoldDershowitz(const double jDay)
 	}
 	else if ((year >= 1800) && (year <= 1986))
 	{
-		double c = (getFixedFromGregorian(1900, 1, 1)-getFixedFromGregorian(year, 7, 1))/36525.;
+		const double c = (getFixedFromGregorian(year, 7, 1)-getFixedFromGregorian(1900, 1, 1))/36525.;
```
Comment (alex-w) [inline src/core/StelUtils.cpp:1779]: Oops… it was my fail :(

**Comment by gzotti** [comment]
> Maybe adding context for calendar names in the GUI will be good idea?

Where do you see necessity? I have added it in `Calendars::draw()` long ago.

File: src/core/StelUtils.cpp (hunk 1 above)
Comment (gzotti) [reply src/core/StelUtils.cpp:1779]: Can happen, but always unfortunate. Finally fixed.

**Comment by gzotti** [comment]
@alex-w I think we can merge this quite large step. Some remaining errors in the astro functions are possible but will only be found as new calendars will come which depend on them.

**Review by alex-w — APPROVED** [review]
Thanks!

**Pull request merged**